### PR TITLE
ENH: Add option in read_csv to infer compression type from filename

### PR DESCRIPTION
**Pull request by evanpw**
Ideally, I would love for this to be the default, but that wouldn't be backwards-compatible in the case where the filename ends in '.gz' or '.bz2' and you want to treat it as uncompressed. That seems like it would be very rare, though.

### CODE REVIEW

File: pandas/parser.pyx [hunk 1]
```diff
@@ -541,6 +541,17 @@ cdef class TextReader:
         self.parser.cb_io = NULL
         self.parser.cb_cleanup = NULL
 
+        if self.compression == 'infer':
+            if isinstance(source, basestring):
+                if source.endswith('.gz'):
+                    self.compression = 'gzip'
+                elif source.endswith('.bz2'):
```
Comment (jreback) [inline pandas/parser.pyx]: better to do an assert_frame_equal herr

**Comment by shoyer** [comment]
I think it would even be fine even to change the default here. We are not *that* strict about backwards compatibility in pandas -- any users who relied on the previous behavior were basically relying on a bug.

**Comment by jreback** [comment]
I agree with @shoyer here, let's just have it infer these filename endings as compression (move the release note to the API section).

**Comment by evanpw** [comment]
I've totally borked this branch with an accidental force push. I'll fix it tonight.

**Comment by evanpw** [comment]
Should be fixed.

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -55,8 +55,10 @@ class ParserWarning(Warning):
 dtype : Type name or dict of column -> type
     Data type for data or columns. E.g. {'a': np.float64, 'b': np.int32}
     (Unsupported with engine='python')
-compression : {'gzip', 'bz2', None}, default None
-    For on-the-fly decompression of on-disk data
+compression : {'gzip', 'bz2', 'infer', None}, default 'infer'
+    For on-the-fly decompression of on-disk data. If 'infer', then use gzip or
+    bz2 if filepath_or_buffer is a string ending in '.gz' or '.bz2',
+    respectively, and no decompression otherwise.
```
Comment (jorisvandenbossche) [inline pandas/io/parsers.py]: Can you add here what None does?

File: doc/source/whatsnew/v0.16.1.txt [hunk 3]
```diff
@@ -55,6 +56,8 @@ API changes
 - :meth:`~pandas.DataFrame.assign` now inserts new columns in alphabetical order. Previously
   the order was arbitrary. (:issue:`9777`)
 
+- By default, ``read_csv`` and ``read_table`` will now try to infer the compression type based on the file extension. Set ``compression=None`` to restore the previous behavior. (:issue:`9770`)
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.16.1.txt]: say what the previous was (e.g. no decompression)

File: pandas/io/tests/test_parsers.py [hunk 4]
```diff
@@ -1098,6 +1098,20 @@ def test_read_csv_no_index_name(self):
         self.assertEqual(df.ix[:, ['A', 'B', 'C', 'D']].values.dtype, np.float64)
         tm.assert_frame_equal(df, df2)
 
+    def test_read_csv_infer_compression(self):
+        expected = self.read_csv(self.csv1, index_col=0, parse_dates=True)
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: add the issue number here as a comment

**Comment by jreback** [comment]
couple of minor comments, pls rebase, and ping when green

**Comment by evanpw** [comment]
Docs are fixed, rebased/squashed, and tests are green.

**Comment by jreback** [comment]
lgtm

@shoyer @jorisvandenbossche

**Comment by shoyer** [comment]
Looks great to me!

**Comment by jreback** [comment]
@evanpw thanks!

**Comment by evanpw** [comment]
Thank you!

**Pull request merged**